test/tbb/test_collaborative_call_once.cpp

```diff
@@ -0,0 +1,372 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#define TBB_PREVIEW_COLLABORATIVE_CALL_ONCE 1
+#if _MSC_VER && !defined(__INTEL_COMPILER)
+// unreachable code
+#pragma warning( push )
+#pragma warning( disable: 4702 )
+#endif
+
+#include "common/test.h"
+#include "common/exception_handling.h"
+#include "common/utils_concurrency_limit.h"
+
+
+#include "tbb/collaborative_call_once.h"
+#include "tbb/parallel_invoke.h"
+#include "tbb/parallel_reduce.h"
+#include "tbb/task_arena.h"
+
+#include <atomic>
+#include <mutex>
+
+//! \file test_collaborative_call_once.cpp
+//! \brief Tests for [algorithms.collaborative_call_once] functionality
+
+struct increment_functor {
+    std::atomic<int> ct{0};
+
+    void operator()() {
+        ct++;
+    }
+};
+
+struct sum_functor {
+    int sum{0};
+
+    template<typename T>
+    void operator()(T first_op) {
+        sum += first_op;
+    }
+
+    template<typename T, typename... Args>
+    void operator()(T first_op, Args&&... args) {
+        (*this)(first_op);
+        (*this)(std::forward<Args>(args)...);
+    }
+};
+
+class call_once_exception : public std::exception {};
+
+template<typename Fn, typename... Args>
+void call_once_in_for_loop(Fn& body, Args&&... args) {
+    tbb::collaborative_once_flag flag;
+    for (int i = 0; i < 1024; ++i) {
+        tbb::collaborative_call_once(flag, body, std::forward<Args>(args)...);
+    }
+}
+
+template<typename Fn, typename... Args>
+void call_once_in_parallel_for(Fn& body, Args&&... args) {
+    tbb::collaborative_once_flag flag;
+#if __GNUC__ && !__TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT
+    auto stored_pack = tbb::detail::d0::save_pack(std::forward<Args>(args)...);
+    auto func = [&] { tbb::detail::d0::call(body, stored_pack); };
+#endif // !__TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT
+
+    tbb::parallel_for(tbb::blocked_range<size_t>(0, 100), [&](const tbb::blocked_range<size_t>& range) {
+        for (size_t i = range.begin(); i != range.end(); ++i) {
+#if __GNUC__ && !__TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT
+            tbb::collaborative_call_once(flag, func);
+#else
+            tbb::collaborative_call_once(flag, body, std::forward<Args>(args)...);
+#endif // __TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT
+        }
+    });
+}
+
+template<typename Fn, typename... Args>
+void call_once_threads(Fn& body, Args&&... args) {
+    tbb::collaborative_once_flag flag;
+    std::vector<std::thread> threads;
+
+#if __GNUC__ && !__TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT
```

Comment:
I set check for `__GNUC__`, because I think, that the general way to test here- it is using forwarding to `collaborative_call_once` directly. `stored_pack` here - is a workaround for GCC 4.8.5 bug: https://godbolt.org/z/YbTM1qfPs

Reply:
Maybe we will rename `__TBB_GCC_PARAMETER_PACK_IN_LAMBDAS_PRESENT ` to `__TBB_PARAMETER_PACK_IN_LAMBDAS_BROKEN`?